### PR TITLE
Fix rustdoc

### LINKED ISSUE #516: Documenting Private Items Fails
When trying to generate all of our API documentation (useful for doing low-level work), the following fails with numerous warnings and errors.

```
$ cargo doc --open --document-private-items
```

While not critical to our public documentation, we probably should be strict and make sure this doesn't break.  I suggest fixing the errors (and probably the warnings) and put this as a step in CI to make sure we keep ourselves honest.

### PR DESCRIPTION
**Pull request by almann**
Cleans up doc comments such that `cargo doc --document-private-items --all-features` works and is in CI.

Fixes #516

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/text/non_blocking/raw_text_reader.rs [hunk 1]
```diff
@@ -404,9 +404,11 @@ impl<A: AsRef<[u8]> + Expandable> RawTextReader<A> {
     }
 
     /// Assumes that the reader is inside a struct AND that a field has already been successfully
-    /// parsed from input using [next_struct_field_name] and attempts to parse the next value.
+    /// parsed from input using [`next_struct_field_name`][1] and attempts to parse the next value.
     /// In this input position, only a value (or whitespace/comments) are legal. Anything else
     /// (including EOF) will result in a decoding error.
+    ///
+    /// [1]: Self::next_struct_field_name
```
Comment (jobarr-amzn) [inline src/text/non_blocking/raw_text_reader.rs:411]: Only two examples in this patch of a reference-style link, and none that I can find outside it. Might as well stick with inline links, and if we find a more complicated case use named reference-style links. 
```suggestion
    /// parsed from input using [`next_struct_field_name`](Self::next_struct_field_name) and
    /// attempts to parse the next value.
    /// In this input position, only a value (or whitespace/comments) are legal. Anything else
    /// (including EOF) will result in a decoding error.
```

File: src/text/parsers/text_support.rs [hunk 2]
```diff
@@ -177,10 +177,12 @@ pub(crate) fn escaped_char_unicode(input: &str) -> IonParseResult<char> {
     Ok((remaining_input, character))
 }
 
-/// Rust's [char] type represents any Unicode code point EXCEPT a surrogate. If we encounter a high
+/// Rust's [`char`][1] type represents any Unicode code point EXCEPT a surrogate. If we encounter a high
 /// surrogate in the stream, we cannot convert it to a `char` yet. We must find the (mandatory)
 /// low surrogate that follows it in the stream; then we can combine the high and low surrogates
```
Comment (jobarr-amzn) [inline src/text/parsers/text_support.rs:182]:
```suggestion
/// Rust's [`char`](prim@char) type represents any Unicode code point EXCEPT a surrogate. If we
/// encounter a high surrogate in the stream, we cannot convert it to a `char` yet. We must find the
/// (mandatory) low surrogate that follows it in the stream; then we can combine the high and low
/// surrogates into a complete code point. This code point can then be returned as a Rust
```

File: src/text/parsers/text_support.rs [hunk 3]
```diff
@@ -177,10 +177,12 @@ pub(crate) fn escaped_char_unicode(input: &str) -> IonParseResult<char> {
...
-/// into a complete code point. This code point can then be returned as a Rust [char].
+/// into a complete code point. This code point can then be returned as a Rust [`char`][1].
+///
+/// [1]: prim@char
```
Comment (jobarr-amzn) [inline src/text/parsers/text_support.rs:185]:
```suggestion
/// [`char`](prim@char).
```

File: src/lib.rs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 #![allow(dead_code)]
 #![deny(rustdoc::broken_intra_doc_links)]
+#![deny(rustdoc::bare_urls)]
```
[the remark this answers is not in the document]
Comment (almann) [reply src/lib.rs:3]: This is saying Rustdoc will _error_ when we make this mistake.

**Review by jobarr-amzn — APPROVED** [review]

**Pull request merged**